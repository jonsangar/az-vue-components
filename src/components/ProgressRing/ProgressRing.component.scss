.dashed-circle-progress {
  position: relative;

  svg{
    transform: rotate(-90deg);
    transform-origin: 50% 50%;

    .progress-ring__circle{
      transition: stroke-dashoffset 0.35s;
    }
  }

}

.textInfo{
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: fit-content;
  height: fit-content;
  margin: auto;

  .progress {
    font-size: font-size(h4);
  }

  .label {
    font-weight: bold;
    text-align: center;
  }
}
