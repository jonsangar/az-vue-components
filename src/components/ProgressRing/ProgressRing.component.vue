<template lang="pug">
.dashed-circle-progress(
  :style="componentDimension")
  svg(
    :height="radius * 2"
    :width="radius * 2")
    defs
      mask(
        :id="`mask-${uid}`"
        :height="radius * 2"
        :width="radius * 2"
        )
        circle(
        class="progress-ring__circle"
        stroke="white"
        fill="transparent"
        :stroke-dasharray="circumference + ' ' + circumference"
        :style="{ strokeDashoffset }"
        :stroke-width="stroke"
        :r="normalizedRadius"
        :cx="radius"
        :cy="radius")


    circle(
      :stroke="color2 ? color2 : color"
      fill="transparent"
      style="opacity: 0.3"
      :stroke-dasharray="dashWeight"
      :stroke-width="stroke"
      :r="normalizedRadius"
      :cx="radius"
      :cy="radius")

    circle(
      :stroke="color"
      fill="transparent"
      :stroke-dasharray="dashWeight"
      :style="`mask: url('#mask-${uid}'); opacity: 1`"
      :stroke-width="stroke"
      :r="normalizedRadius"
      :cx="radius"
      :cy="radius")

  .textInfo(
    :style="textStyle")

    .progress(v-if="showProgress")
      | {{ progress }}
    .label(v-else-if="label")
      | {{ label }}

</template>

<script>
export default {
  props: {
    radius: {
      type: Number,
      required: true
    },
    progress: {
      type: Number,
      required: true
    },
    stroke: {
      type: Number,
      required: true
    },
    color: {
      type: String,
      required: true
    },
    color2: {
      type: String,
      required: false
    },
    showProgress: {
      type: Boolean,
      default: false
    },
    dashWeight: {
      type: Number,
      default: 2
    },
    label: {
      type: String,
      required: false
    }
  },
  data() {
    const normalizedRadius = this.radius - this.stroke * 2;
    const circumference = normalizedRadius * 2 * Math.PI;
    return {
      normalizedRadius,
      circumference,
      uid: this._uid
    }
  },

  computed: {
    componentDimension() {
      return {
        width: `${this.radius * 2}px`,
        height: `${this.radius * 2}px`
      }
    },

    textStyle() {
      return {
        color: this.color,
        maxWidth: `${this.radius}px`
      }
    },

    strokeDashoffset() {
      return this.circumference - this.progress / 100 * this.circumference;
    }
  },

  created() {
    this.uid = this._uid
  },

}
</script>

<style lang="scss" src="./ProgressRing.component.scss" scoped></style>
