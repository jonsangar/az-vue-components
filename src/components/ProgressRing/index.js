import ProgressRing from './ProgressRing.component.vue';

export default ProgressRing;
