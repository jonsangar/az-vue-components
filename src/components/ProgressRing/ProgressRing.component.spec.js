import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerÃ­as de estilos e iconos
import SvgIcon from 'vue-svgicon'
import ProgressRing from './ProgressRing.component.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('ProgressRing.vue', () => {
  const build = () => {
    const wrapper = shallowMount(ProgressRing, {
      localVue,
      propsData: {
        radius: 5,
        progress: 10,
        stroke: 1,
        color: 'light',
        color2: 'dark',
        showProgress: true,
        dashWeight: 2,
        label: 'Cargando'
      },
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente ProgressRing con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.vm.radius).toBe(5)
    expect(wrapper.vm.progress).toBe(10)
    expect(wrapper.vm.stroke).toBe(1)
    expect(wrapper.vm.color).toBe('light')
  })

})

describe('ProgressRing.vue', () => {
  const build = () => {
    const wrapper = shallowMount(ProgressRing, {
      localVue,
      propsData: {
        radius: 10,
        progress: 100,
        stroke: 2,
        color: 'blue',
        color2: 'dark',
        showProgress: true,
        dashWeight: 2,
        label: 'Cargando'
      },
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que se renderiza las props cuando se asignan las props', async () => {
    const { wrapper } = build()

    expect(wrapper.vm.radius).toBe(10)
    expect(wrapper.vm.progress).toBe(100)
    expect(wrapper.vm.stroke).toBe(2)
    expect(wrapper.vm.color).toBe('blue')
  })

});
