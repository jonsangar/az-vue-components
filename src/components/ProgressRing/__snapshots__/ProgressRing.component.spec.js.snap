// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressRing.vue Comprueba renderizado del componente ProgressRing con snapshot 1`] = `
<div class="dashed-circle-progress" style="width: 10px; height: 10px;"><svg height="10" width="10">
    <defs>
      <mask id="mask-1" height="10" width="10">
        <circle stroke="white" fill="transparent" stroke-dasharray="18.84955592153876 18.84955592153876" stroke-width="1" r="3" cx="5" cy="5" class="progress-ring__circle" style="stroke-dashoffset: 16.964600329384883;"></circle>
      </mask>
    </defs>
    <circle stroke="dark" fill="transparent" stroke-dasharray="2" stroke-width="1" r="3" cx="5" cy="5" style="opacity: 0.3;"></circle>
    <circle stroke="light" fill="transparent" stroke-dasharray="2" stroke-width="1" r="3" cx="5" cy="5" style="mask: url('#mask-1'); opacity: 1;"></circle>
  </svg>
  <div class="textInfo" style="max-width: 5px;">
    <div class="progress">10</div>
  </div>
</div>
`;
